stdlib/public/SwiftShims/RefCount.h

```diff
@@ -20,24 +20,503 @@
+    becomes DEINITING.
+
+  LIVE with side table
+  Weak variable operations work normally.
+  Everything else is the same as LIVE.
+
+  DEINITING without side table
+  deinit() is in progress on the object.
+  Strong variable operations have no effect.
+  Unowned variable load halts in swift_abortRetainUnowned().
+  Unowned variable store works normally.
+  Weak variable load can't happen.
+  Weak variable store stores nil.
+  When deinit() completes, the generated code calls swift_deallocObject. 
+    swift_deallocObject calls canBeFreedNow() checking for the fast path 
+    of no weak or unowned references. 
+    If canBeFreedNow() the object is freed and it becomes DEAD. 
+    Otherwise, it decrements the unowned RC and the object becomes DEINITED.
+
+  DEINITING with side table
+  Weak variable load returns nil. 
+  Weak variable store stores nil.
+  canBeFreedNow() is always false, so it never transitions directly to DEAD.
+  Everything else is the same as DEINITING.
+
+  DEINITED without side table
+  deinit() has completed but there are unowned references outstanding.
+  Strong variable operations can't happen.
+  Unowned variable store can't happen.
+  Unowned variable load halts in swift_abortRetainUnowned().
+  Weak variable operations can't happen.
+  When the unowned RC reaches zero, the object is freed and it becomes DEAD.
+
+  DEINITED with side table
+  Weak variable load returns nil.
+  Weak variable store can't happen.
+  When the unowned RC reaches zero, the object is freed, the weak RC is 
+    decremented, and the object becomes FREED.
+  Everything else is the same as DEINITED.
+
+  FREED without side table
+  This state never happens.
+
+  FREED with side table
+  The object is freed but there are weak refs to the side table outstanding.
+  Strong variable operations can't happen.
+  Unowned variable operations can't happen.
+  Weak variable load returns nil.
+  Weak variable store can't happen.
+  When the weak RC reaches zero, the side table entry is freed and 
+    the object becomes DEAD.
+
+  DEAD
+  The object and its side table are gone.
+*/
+
+namespace swift {
+  struct HeapObject;
+  class HeapObjectSideTableEntry;
+}
+
+// FIXME: HACK: copied from HeapObject.cpp
+extern "C" LLVM_LIBRARY_VISIBILITY void
+_swift_release_dealloc(swift::HeapObject *object)
+  SWIFT_CC(RegisterPreservingCC_IMPL)
+    __attribute__((__noinline__, __used__));
+
+namespace swift {
+
+// FIXME: many `relaxed` in this file should be `consume`,
+// but (1) the compiler doesn't support `consume` directly,
+// and (2) the compiler promotes `consume` to `acquire` instead which
+// is overkill on our CPUs. But this might leave us vulnerable to
+// compiler optimizations that `relaxed` allows but `consume` ought not.
+#define relaxed std::memory_order_relaxed
+#define acquire std::memory_order_acquire
+#define release std::memory_order_release
+#define consume std::memory_order_consume
```

Comment:
Are these macros really worth it?  You don't have *that* many uses of them, and I actually like the clarity of seeing them written out.